[CI] EsqlSpecIT test {string.ReverseGraphemeClusters ASYNC} failing

**Build Scans:**
- [elasticsearch-periodic #4367 / graalvm-ce17_checkpart3_java-matrix](https://gradle-enterprise.elastic.co/s/hi2kajpqi5bcs)
- [elasticsearch-periodic #4367 / openjdk17_checkpart3_java-matrix](https://gradle-enterprise.elastic.co/s/f5e7whta3swse)
- [elasticsearch-periodic #4367 / openjdk17_checkpart3_java-fips-matrix](https://gradle-enterprise.elastic.co/s/hgtv5hhzipuc4)
- [elasticsearch-periodic #4363 / graalvm-ce17_checkpart3_java-matrix](https://gradle-enterprise.elastic.co/s/cvzdm3tkbkns6)
- [elasticsearch-periodic #4363 / openjdk17_checkpart3_java-fips-matrix](https://gradle-enterprise.elastic.co/s/mju2gkje2lpri)
- [elasticsearch-periodic #4363 / openjdk17_checkpart3_java-matrix](https://gradle-enterprise.elastic.co/s/xjeizyk6iwyzg)
- [elasticsearch-periodic #4359 / openjdk17_checkpart3_java-matrix](https://gradle-enterprise.elastic.co/s/ypdzwjdpcdegc)
- [elasticsearch-periodic #4359 / openjdk17_checkpart3_java-fips-matrix](https://gradle-enterprise.elastic.co/s/w42xs3s54hau2)
- [elasticsearch-periodic #4359 / graalvm-ce17_checkpart3_java-matrix](https://gradle-enterprise.elastic.co/s/u7dfrjrkvpen2)
- [elasticsearch-periodic #4355 / graalvm-ce17_checkpart3_java-matrix](https://gradle-enterprise.elastic.co/s/q3eaiz6qt3mbs)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:qa:server:multi-node:javaRestTest" --tests "org.elasticsearch.xpack.esql.qa.multi_node.EsqlSpecIT" -Dtests.method="test {string.ReverseGraphemeClusters ASYNC}" -Dtests.seed=3F37804BD9C05A68 -Dtests.locale=mk-Cyrl-MK -Dtests.timezone=America/Iqaluit -Druntime.java=17
```

**Applicable branches:**
8.x

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('test%20%7Bstring.ReverseGraphemeClusters%20ASYNC%7D'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.esql.qa.multi_node.EsqlSpecIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: Data mismatch:
row 0 column 1:a list containing
row 0 column 1:0: expected "ठाीकं💖🎉👍🏽😊ûôîêâùòìèàúóíéá" but was "ीठाकं💖🎉🏽👍😊ûôîêâùòìèàúóíéá"
```

**Issue Reasons:**
- [8.x] 17 consecutive failures in step graalvm-ce17_checkpart3_java-matrix
- [8.x] 17 consecutive failures in step openjdk17_checkpart3_java-matrix
- [8.x] 17 consecutive failures in step openjdk17_checkpart3_java-fips-matrix
- [8.x] 51 failures in test test {string.ReverseGraphemeClusters ASYNC} (8.7% fail rate in 586 executions)
- [8.x] 17 failures in step graalvm-ce17_checkpart3_java-matrix (100.0% fail rate in 17 executions)
- [8.x] 17 failures in step openjdk17_checkpart3_java-matrix (100.0% fail rate in 17 executions)
- [8.x] 17 failures in step openjdk17_checkpart3_java-fips-matrix (100.0% fail rate in 17 executions)
- [8.x] 17 failures in pipeline elasticsearch-periodic (100.0% fail rate in 17 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/docs/reference/esql/functions/description/reverse.asciidoc
+++ b/docs/reference/esql/functions/description/reverse.asciidoc
@@ -3,3 +3,7 @@
 *Description*
 
 Returns a new string representing the input string in reverse order.
+
+NOTE: If Elasticsearch is running with a JDK version less than 20 then this will not properly reverse Grapheme Clusters.
+Elastic Cloud the JDK bundled with Elasticsearch all use newer JDKs. But if you've explicitly shifted to an older jdk
+then you'll see things like "👍🏽😊" be reversed to  "🏽👍😊" instead of the correct "😊👍🏽".

--- a/docs/reference/esql/functions/kibana/definition/reverse.json
+++ b/docs/reference/esql/functions/kibana/definition/reverse.json
@@ -3,6 +3,7 @@
   "type" : "eval",
   "name" : "reverse",
   "description" : "Returns a new string representing the input string in reverse order.",
+  "note" : "If Elasticsearch is running with a JDK version less than 20 then this will not properly reverse Grapheme Clusters.\nElastic Cloud the JDK bundled with Elasticsearch all use newer JDKs. But if you've explicitly shifted to an older jdk\nthen you'll see things like \"\uD83D\uDC4D\uD83C\uDFFD\uD83D\uDE0A\" be reversed to  \"\uD83C\uDFFD\uD83D\uDC4D\uD83D\uDE0A\" instead of the correct \"\uD83D\uDE0A\uD83D\uDC4D\uD83C\uDFFD\".",
   "signatures" : [
     {
       "params" : [

--- a/docs/reference/esql/functions/kibana/docs/reverse.md
+++ b/docs/reference/esql/functions/kibana/docs/reverse.md
@@ -8,3 +8,6 @@ Returns a new string representing the input string in reverse order.
 ```
 ROW message = "Some Text" | EVAL message_reversed = REVERSE(message);
 ```
+Note: If Elasticsearch is running with a JDK version less than 20 then this will not properly reverse Grapheme Clusters.
+Elastic Cloud the JDK bundled with Elasticsearch all use newer JDKs. But if you've explicitly shifted to an older jdk
+then you'll see things like "👍🏽😊" be reversed to  "🏽👍😊" instead of the correct "😊👍🏽".

--- a/x-pack/plugin/esql/qa/testFixtures/src/main/resources/string.csv-spec
+++ b/x-pack/plugin/esql/qa/testFixtures/src/main/resources/string.csv-spec
@@ -1236,6 +1236,7 @@ off_on_holiday:keyword | back_home_again:keyword
 
 reverseGraphemeClusters
 required_capability: fn_reverse
+required_capability: fn_reverse_grapheme_clusters
 ROW message = "áéíóúàèìòùâêîôû😊👍🏽🎉💖कंठाी" | EVAL message_reversed = REVERSE(message);
 
 message:keyword                | message_reversed:keyword

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java
@@ -33,6 +33,12 @@ public class EsqlCapabilities {
         FN_REVERSE,
 
         /**
+         * Support for reversing whole grapheme clusters. This is not supported
+         * on JDK versions less than 20.
+         */
+        FN_REVERSE_GRAPHEME_CLUSTERS(Runtime.version().feature() < 20),
+
+        /**
          * Support for function {@code CBRT}. Done in #108574.
          */
         FN_CBRT,
@@ -133,7 +139,7 @@ public class EsqlCapabilities {
          * - fixed variable shadowing
          * - fixed Join.references(), requiring breaking change to Join serialization
          */
-        LOOKUP_V4(true),
+        LOOKUP_V4(Build.current().isSnapshot()),
 
         /**
          * Support for requesting the "REPEAT" command.
@@ -279,7 +285,7 @@ public class EsqlCapabilities {
         /**
          * Support for match operator
          */
-        MATCH_OPERATOR(true),
+        MATCH_OPERATOR(Build.current().isSnapshot()),
 
         /**
          * Removing support for the {@code META} keyword.
@@ -349,7 +355,7 @@ public class EsqlCapabilities {
         /**
          * Supported the text categorization function "CATEGORIZE".
          */
-        CATEGORIZE(true),
+        CATEGORIZE(Build.current().isSnapshot()),
 
         /**
          * QSTR function
@@ -375,7 +381,7 @@ public class EsqlCapabilities {
         /**
          * Support named parameters for field names.
          */
-        NAMED_PARAMETER_FOR_FIELD_AND_FUNCTION_NAMES(true),
+        NAMED_PARAMETER_FOR_FIELD_AND_FUNCTION_NAMES(Build.current().isSnapshot()),
 
         /**
          * Fix sorting not allowed on _source and counters.
@@ -397,32 +403,22 @@ public class EsqlCapabilities {
          */
         FUNCTION_STATS;
 
-        private final boolean snapshotOnly;
-        private final FeatureFlag featureFlag;
+        private final boolean enabled;
 
         Cap() {
-            this(false, null);
+            this.enabled = true;
         };
 
-        Cap(boolean snapshotOnly) {
-            this(snapshotOnly, null);
+        Cap(boolean enabled) {
+            this.enabled = enabled;
         };
 
         Cap(FeatureFlag featureFlag) {
-            this(false, featureFlag);
-        }
-
-        Cap(boolean snapshotOnly, FeatureFlag featureFlag) {
-            assert featureFlag == null || snapshotOnly == false;
-            this.snapshotOnly = snapshotOnly;
-            this.featureFlag = featureFlag;
+            this.enabled = featureFlag.isEnabled();
         }
 
         public boolean isEnabled() {
-            if (featureFlag == null) {
-                return Build.current().isSnapshot() || this.snapshotOnly == false;
-            }
-            return featureFlag.isEnabled();
+            return enabled;
         }
 
         public String capabilityName() {
@@ -430,12 +426,17 @@ public class EsqlCapabilities {
         }
     }
 
-    public static final Set<String> CAPABILITIES = capabilities();
+    public static final Set<String> CAPABILITIES = capabilities(false);
 
-    private static Set<String> capabilities() {
+    /**
+     * Get a {@link Set} of all capabilities. If the {@code all} parameter is {@code false}
+     * then only <strong>enabled</strong> capabilities are returned - otherwise <strong>all</strong>
+     * known capabilities are returned.
+     */
+    public static Set<String> capabilities(boolean all) {
         List<String> caps = new ArrayList<>();
         for (Cap cap : Cap.values()) {
-            if (cap.isEnabled()) {
+            if (all || cap.isEnabled()) {
                 caps.add(cap.capabilityName());
             }
         }

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Reverse.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Reverse.java
@@ -49,7 +49,7 @@ public class Reverse extends UnaryScalarFunction {
             ) },
         note = """
             If Elasticsearch is running with a JDK version less than 20 then this will not properly reverse Grapheme Clusters.
-            Elastic Cloud the JDK bundled with Elasticsearch all use newer JDKs. But if you've explicitly shifted to an older jdk
+            Elastic Cloud and the JDK bundled with Elasticsearch all use newer JDKs. But if you've explicitly shifted to an older jdk
             then you'll see things like "👍🏽😊" be reversed to  "🏽👍😊" instead of the correct "😊👍🏽"."""
     )
     public Reverse(

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Reverse.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Reverse.java
@@ -46,7 +46,11 @@ public class Reverse extends UnaryScalarFunction {
                 file = "string",
                 tag = "reverseEmoji",
                 description = "`REVERSE` works with unicode, too! It keeps unicode grapheme clusters together during reversal."
-            ) }
+            ) },
+        note = """
+            If Elasticsearch is running with a JDK version less than 20 then this will not properly reverse Grapheme Clusters.
+            Elastic Cloud the JDK bundled with Elasticsearch all use newer JDKs. But if you've explicitly shifted to an older jdk
+            then you'll see things like "👍🏽😊" be reversed to  "🏽👍😊" instead of the correct "😊👍🏽"."""
     )
     public Reverse(
         Source source,

--- a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/CsvTests.java
+++ b/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/CsvTests.java
@@ -257,7 +257,7 @@ public class CsvTests extends ESTestCase {
                 assertThat(
                     "Capability is not included in the enabled list capabilities on a snapshot build. Spelling mistake?",
                     testCase.requiredCapabilities,
-                    everyItem(in(EsqlCapabilities.CAPABILITIES))
+                    everyItem(in(EsqlCapabilities.capabilities(true)))
                 );
             } else {
                 for (EsqlCapabilities.Cap c : EsqlCapabilities.Cap.values()) {

